reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+                                    pullIntoDescriptor.byteLength - pullIntoDescriptor.bytesFilled]);
+  } catch (e) {
+    DestroyReadableByteStreamController(controller);
+    const stream = controller._controlledReadableByteStream;
+    if (stream._state === 'readable') {
+      ErrorReadableByteStream(stream, e);
+    }
+  }
+
+  controller._insideUnderlyingByteSource = false;
+}
+
+function CallPullOrPullIntoLaterIfNeeded(controller) {
+  if (controller._insideUnderlyingByteSource) {
+    controller._callPullOrPullIntoLaterIfNeeded = true;
+    return;
+  }
+
+  Promise.resolve().then(CallPullOrPullIntoRepeatedlyIfNeeded.bind(undefined, controller));
+}
+
+function CallPullOrPullIntoRepeatedlyIfNeeded(controller) {
+  const stream = controller._controlledReadableByteStream;
+
+  while (true) {
+    if (!controller._callPullOrPullIntoLaterIfNeeded) {
+      return;
+    }
+
+    if (controller._closeRequested) {
+      return;
+    }
+    if (stream._state !== 'readable') {
+      return;
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined) {
+      return;
+    }
+
+    if (IsReadableByteStreamReader(reader)) {
+      if (reader._readRequests.length === 0) {
+        return;
+      }
+
+      CallPull(controller);
+    } else {
+      assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+      if (reader._readIntoRequests.length === 0) {
+        return;
+      }
+
+      CallPullInto(controller);
+    }
+  }
+}
+
+function CancelReadableByteStream(stream, reason) {
+  if (stream._state === 'closed') {
+    return Promise.resolve(undefined);
+  }
+  if (stream._state === 'errored') {
+    return Promise.reject(stream._storedError);
+  }
+
+  CloseReadableByteStream(stream);
+
+  const controller = stream._controller;
+
+  controller._totalQueuedBytes = 0;
+  controller._queue = [];
+
+  const sourceCancelPromise = PromiseInvokeOrNoop(controller._underlyingByteSource, 'cancel', [reason]);
+  return sourceCancelPromise.then(() => undefined);
+}
+
+function CloseReadableByteStream(stream) {
+  assert(IsReadableByteStream(stream), 'stream must be ReadableByteStream');
+  assert(stream._state === 'readable', 'state must be readable');
+
+  stream._state = 'closed';
+
+  if (IsReadableByteStreamLocked(stream)) {
+    CloseReadableByteStreamReader(stream._reader);
+  }
+}
+
+function CloseReadableByteStreamReader(reader) {
+  reader._state = 'closed';
+
+  reader._closedPromise_resolve(undefined);
+  reader._closedPromise_resolve = undefined;
+  reader._closedPromise_reject = undefined;
+
+  if (IsReadableByteStreamReader(reader)) {
+    for (const req of reader._readRequests) {
+      req.resolve(CreateIterResultObject(undefined, true));
+    }
+
+    reader._readRequests = [];
+  } else {
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    if (reader._readIntoRequests.length > 0) {
+      return;
```

Comment:
Yeah, this is complicated.

We have 3 options:

1. Wait for the buffer to return. Once returned, `controller.respond(0, returnedBuffer)` and `controller.close()`
1. `controller.close()` and wait for the buffer to return. Once returned `controller.respond(0, returnedBuffer)`
1. Accept the same sequence (at controller side) as (2) but make the reader look not closed until the `respond()` is made

I chose (2) to thinking of a case like this:

At the time when an underlying byte source notices that no more data comes (or on `reader.cancel()` call), the buffer passed via `pullInto()` may still owned by some other thread and we need to wait for the buffer to be returned. (*)

(2) allows faster delivery of closure signal to the consumer.

(1) and (3) are equivalent from the view point of the consumer. So, I'd just choose (1) from them.

I'm not sure (*) is so realistic. But it seems `reader.cancel()` should be able to mark the stream closed synchronously. This requires the same state transition sequence as (2).